control-plane/csi-controller/src/controller.rs

```diff
@@ -0,0 +1,726 @@
+                    "I/O timeout is valid only for nvmf protocol",
+                ));
+            }
+            if io_timeout.parse::<u64>().is_err() {
+                return Err(Status::invalid_argument("Invalid I/O timeout"));
+            }
+        }
+
+        let replica_count: u8 = match args.parameters.get("repl") {
+            Some(c) => match c.parse::<u8>() {
+                Ok(c) => {
+                    if c == 0 {
+                        return Err(Status::invalid_argument(
+                            "Replica count must be greater than zero",
+                        ));
+                    }
+                    c
+                }
+                Err(_) => return Err(Status::invalid_argument("Invalid replica count")),
+            },
+            None => 1,
+        };
+
+        let pinned_volume =
+            volume_opts::decode_local_volume_flag(args.parameters.get(volume_opts::LOCAL_VOLUME));
+
+        // For exaplanation of accessibilityRequirements refer to a table at
+        // https://github.com/kubernetes-csi/external-provisioner.
+        // Our case is WaitForFirstConsumer = true, strict-topology = false.
+        //
+        // The first node in preferred array the node that was chosen for running
+        // the app by the k8s scheduler. The rest of the entries are in random
+        // order and perhaps don't even run mayastor csi node plugin.
+        //
+        // The requisite array contains all nodes in the cluster irrespective
+        // of what node was chosen for running the app.
+        let mut allowed_nodes: Vec<String> = Vec::new();
+        let mut preferred_nodes: Vec<String> = Vec::new();
+
+        if let Some(reqs) = args.accessibility_requirements {
+            for r in reqs.requisite.iter() {
+                for (k, v) in r.segments.iter() {
+                    // We are not able to evaluate any other topology requirements than
+                    // the hostname req. Reject all others.
+                    if k != K8S_HOSTNAME {
+                        return Err(Status::invalid_argument(
+                            "Volume topology other than hostname not supported",
+                        ));
+                    }
+                    allowed_nodes.push(v.to_string());
+                }
+            }
+
+            for p in reqs.preferred.iter() {
+                for (k, v) in p.segments.iter() {
+                    // Ignore others than hostname (it's only preferred)
+                    if k == K8S_HOSTNAME {
+                        preferred_nodes.push(v.to_string());
+                    }
+                }
+            }
+        }
+
+        let u = Uuid::parse_str(&volume_uuid).map_err(|_e| {
+            Status::invalid_argument(format!("Malformed volume UUID: {}", volume_uuid))
+        })?;
+
+        // First check if the volume already exists.
+        if let Some(existing_volume) = MayastorApiClient::get_client()
+            .list_volumes()
```

Comment:
This can list usize:max volumes... I've already mentioned that in the previous comment. Having that said, why do we not get the volume directly? i.e GET volume/{ID} ?

Reply:
GETting volume directly by name will result in 404 every time create_volume() is called (in most of the cases, where we create a new volume), so we will see lots of false-positive errors in REST server logs.
Ideally, we could use GET /volumes?uuid=volume_id and get either 0 elements or 1.
Maybe we can add search options later to REST API.